<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>ete2.phylomedb.phylomeDB.PhylomeDBConnector</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="ete2-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >ete2</th>
          </tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="ete2-module.html">Package&nbsp;ete2</a> ::
        <a href="ete2.phylomedb-module.html">Package&nbsp;phylomedb</a> ::
        <a href="ete2.phylomedb.phylomeDB-module.html">Module&nbsp;phylomeDB</a> ::
        Class&nbsp;PhylomeDBConnector
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript:void(0);" class="privatelink"
    onclick="toggle_private();">hide&nbsp;private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="ete2.phylomedb.phylomeDB.PhylomeDBConnector-class.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== CLASS DESCRIPTION ==================== -->
<h1 class="epydoc">Class PhylomeDBConnector</h1><p class="nomargin-top"><span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector">source&nbsp;code</a></span></p>
<pre class="base-tree">
object --+
         |
        <strong class="uidshort">PhylomeDBConnector</strong>
</pre>

<hr />
<p>Reuturns a connector to a phylomeDB database.</p>
  <h1 class="heading">ARGUMENTS:</h1>
    <p>host: hostname in which phylomeDB is hosted. user: username to the 
    database. passwd: password to connect database. port: port used to 
    connect database.</p>
  <h1 class="heading">RETURNS:</h1>
    <p>An object whose methods can be used to query the database.</p>

<!-- ==================== INSTANCE METHODS ==================== -->
<a name="section-InstanceMethods"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Instance Methods</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-InstanceMethods"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ete2.phylomedb.phylomeDB.PhylomeDBConnector-class.html#__init__" class="summary-sig-name">__init__</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">host</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">84.88.66.245</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">db</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">phylomeDB</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">user</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">public</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">passwd</span>=<span class="summary-sig-default"><code class="variable-quote">'</code><code class="variable-string">public</code><code class="variable-quote">'</code></span>,
        <span class="summary-sig-arg">port</span>=<span class="summary-sig-default">3306</span>)</span><br />
      Connects to a phylomeDB database and returns an object to perform 
      custom queries on it.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.__init__">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr class="private">
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="_execute_block"></a><span class="summary-sig-name">_execute_block</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">cmd</span>)</span><br />
      Executes a multi-line SQL command and returns the nombre of affected 
      rows.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector._execute_block">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_longest_isoform"></a><span class="summary-sig-name">get_longest_isoform</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">phyID</span>)</span><br />
      returns the protID of the</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_longest_isoform">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_id_by_external"></a><span class="summary-sig-name">get_id_by_external</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">external</span>)</span><br />
      Returns the phylomeID of the given external ID</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_id_by_external">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_id_translations"></a><span class="summary-sig-name">get_id_translations</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">seqid</span>)</span><br />
      returns all the registered translations of a given seqid</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_id_translations">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ete2.phylomedb.phylomeDB.PhylomeDBConnector-class.html#search_id" class="summary-sig-name">search_id</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">queryID</span>)</span><br />
      Returns a list of phylome protein Ids associated to the given 
      external queryID.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.search_id">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_proteomes"></a><span class="summary-sig-name">get_proteomes</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Returns all current available proteomes</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_proteomes">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_species"></a><span class="summary-sig-name">get_species</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Returns all current available species</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_species">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_phylomes"></a><span class="summary-sig-name">get_phylomes</span>(<span class="summary-sig-arg">self</span>)</span><br />
      Returns all current available phylomes</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_phylomes">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_proteomes_in_phylome"></a><span class="summary-sig-name">get_proteomes_in_phylome</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">phylome_id</span>)</span><br />
      Returns a list of proteomes associated to a given phylome_id</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_proteomes_in_phylome">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_seqids_in_proteome"></a><span class="summary-sig-name">get_seqids_in_proteome</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">proteome_id</span>,
        <span class="summary-sig-arg">filter_isoforms</span>=<span class="summary-sig-default">True</span>)</span><br />
      Returns all sequences of a given proteome</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_seqids_in_proteome">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_seqs_in_proteome"></a><span class="summary-sig-name">get_seqs_in_proteome</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">proteome_id</span>,
        <span class="summary-sig-arg">filter_isoforms</span>=<span class="summary-sig-default">True</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_seqs_in_proteome">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_proteome_info"></a><span class="summary-sig-name">get_proteome_info</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">proteome_id</span>)</span><br />
      Returns all info about a registered proteome</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_proteome_info">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_seqid_info"></a><span class="summary-sig-name">get_seqid_info</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">protid</span>)</span><br />
      Returns orginal info about a given protid</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_seqid_info">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_phylome_info"></a><span class="summary-sig-name">get_phylome_info</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">phylomeid</span>)</span><br />
      Returns info on a given phylome</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_phylome_info">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_species_info"></a><span class="summary-sig-name">get_species_info</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">taxid_or_code</span>)</span><br />
      Returns all information on a given species_code</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_species_info">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_seed_ids"></a><span class="summary-sig-name">get_seed_ids</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">phylome_id</span>,
        <span class="summary-sig-arg">filter_isoforms</span>=<span class="summary-sig-default">True</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_seed_ids">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_collateral_seeds"></a><span class="summary-sig-name">get_collateral_seeds</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">seqid</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_collateral_seeds">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_available_trees"></a><span class="summary-sig-name">get_available_trees</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">seqid</span>,
        <span class="summary-sig-arg">collateral</span>=<span class="summary-sig-default">True</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_available_trees">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_available_trees_by_phylome"></a><span class="summary-sig-name">get_available_trees_by_phylome</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">seqid</span>,
        <span class="summary-sig-arg">collateral</span>=<span class="summary-sig-default">True</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_available_trees_by_phylome">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_available_trees_in_phylome"></a><span class="summary-sig-name">get_available_trees_in_phylome</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">seqid</span>,
        <span class="summary-sig-arg">phylomeid</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_available_trees_in_phylome">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_tree"></a><span class="summary-sig-name">get_tree</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">protid</span>,
        <span class="summary-sig-arg">method</span>,
        <span class="summary-sig-arg">phylome_id</span>)</span><br />
      Returns the method-tree associated to a given protid.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_tree">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_best_tree"></a><span class="summary-sig-name">get_best_tree</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">protid</span>,
        <span class="summary-sig-arg">phylome_id</span>)</span><br />
      Returns the winner ML tree</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_best_tree">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_algs"></a><span class="summary-sig-name">get_algs</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">protid</span>,
        <span class="summary-sig-arg">phylome_id</span>)</span><br />
      Given a protID, it returns a tuple with the raw_alg, clean_alg and 
      the number of seqs included.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_algs">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_raw_alg"></a><span class="summary-sig-name">get_raw_alg</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">protid</span>,
        <span class="summary-sig-arg">phylome_id</span>)</span><br />
      Given a protID, it returns a tuple with the raw_alg and the number of
      seqs included.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_raw_alg">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_clean_alg"></a><span class="summary-sig-name">get_clean_alg</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">protid</span>,
        <span class="summary-sig-arg">phylome_id</span>)</span><br />
      Given a protID, it returns a tuple with the clean_alg and the number 
      of seqs included.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_clean_alg">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_phylome_trees"></a><span class="summary-sig-name">get_phylome_trees</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">phylomeid</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_phylome_trees">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="get_phylome_algs"></a><span class="summary-sig-name">get_phylome_algs</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">phylomeid</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_phylome_algs">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="count_trees"></a><span class="summary-sig-name">count_trees</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">phylomeid</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.count_trees">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a name="count_algs"></a><span class="summary-sig-name">count_algs</span>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">phylomeid</span>)</span></td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.count_algs">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
<tr>
    <td width="15%" align="right" valign="top" class="summary">
      <span class="summary-type">&nbsp;</span>
    </td><td class="summary">
      <table width="100%" cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td><span class="summary-sig"><a href="ete2.phylomedb.phylomeDB.PhylomeDBConnector-class.html#get_orthologs" class="summary-sig-name">get_orthologs</a>(<span class="summary-sig-arg">self</span>,
        <span class="summary-sig-arg">seqid</span>,
        <span class="summary-sig-arg">sp2age</span>=<span class="summary-sig-default">None</span>)</span><br />
      Returns the orthology predictions of the given seqid in all
phylomes.</td>
          <td align="right" valign="top">
            <span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_orthologs">source&nbsp;code</a></span>
            
          </td>
        </tr>
      </table>
      
    </td>
  </tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__delattr__</code>,
      <code>__format__</code>,
      <code>__getattribute__</code>,
      <code>__hash__</code>,
      <code>__new__</code>,
      <code>__reduce__</code>,
      <code>__reduce_ex__</code>,
      <code>__repr__</code>,
      <code>__setattr__</code>,
      <code>__sizeof__</code>,
      <code>__str__</code>,
      <code>__subclasshook__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== PROPERTIES ==================== -->
<a name="section-Properties"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Properties</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Properties"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
  <tr>
    <td colspan="2" class="summary">
    <p class="indent-wrapped-lines"><b>Inherited from <code>object</code></b>:
      <code>__class__</code>
      </p>
    </td>
  </tr>
</table>
<!-- ==================== METHOD DETAILS ==================== -->
<a name="section-MethodDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="table-header">
  <td colspan="2" class="table-header">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><span class="table-header">Method Details</span></td>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-MethodDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="__init__"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">__init__</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">host</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">84.88.66.245</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">db</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">phylomeDB</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">user</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">public</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">passwd</span>=<span class="sig-default"><code class="variable-quote">'</code><code class="variable-string">public</code><code class="variable-quote">'</code></span>,
        <span class="sig-arg">port</span>=<span class="sig-default">3306</span>)</span>
    <br /><em class="fname">(Constructor)</em>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.__init__">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Connects to a phylomeDB database and returns an object to perform 
  custom queries on it.</p>
  <dl class="fields">
    <dt>Overrides:
        object.__init__
    </dt>
  </dl>
</td></tr></table>
</div>
<a name="search_id"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">search_id</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">queryID</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.search_id">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <p>Returns a list of phylome protein Ids associated to the given external
  queryID. If queryID is a phylomeDB id, it returns the longest isoform 
  associated to the queryID's gene</p>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<a name="get_orthologs"></a>
<div>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3 class="epydoc"><span class="sig"><span class="sig-name">get_orthologs</span>(<span class="sig-arg">self</span>,
        <span class="sig-arg">seqid</span>,
        <span class="sig-arg">sp2age</span>=<span class="sig-default">None</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="ete2.phylomedb.phylomeDB-pysrc.html#PhylomeDBConnector.get_orthologs">source&nbsp;code</a></span>&nbsp;
    </td>
  </tr></table>
  
  <pre class="literalblock">
Returns the orthology predictions of the given seqid in all
phylomes.

Only seed trees will be used to detect orthologies, and trees will
be rooted as the default policy defined in the API. If phylome has
an asociated dictionary of species ages,
root_to_farthest_oldest_leaf algorithm will be applied. Otherwise,
midpoint is used.

You can also provide your own species age dictionary to force the
rooting of the trees according to such data.


ARGUMENTS:
==========

 seqid: the ID of a sequence in the phylomeDB format.
  i.e. Hsa0000001

 sp2age: a dictionary of species code ages (key=species_code,
  value=age).  i.e. {&quot;Hsa&quot;:1, &quot;Dme&quot;:4, &quot;Ath&quot;:10}

RETURNS:
=========

A dictionary of orthologs and inparalogs found in each scanned
phylomes.

 

</pre>
  <dl class="fields">
  </dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="ete2-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="module-tree.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th>&nbsp;&nbsp;&nbsp;<a
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center"
            >ete2</th>
          </tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">
    Generated by Epydoc 3.0.1 on Fri Jul  2 11:17:25 2010
    </td>
    <td align="right" class="footer">
      <a target="mainFrame" href="http://epydoc.sourceforge.net"
        >http://epydoc.sourceforge.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie();
  // -->
</script>
</body>
</html>
