diff -u -d -r mod_python-3.3.1-orig/src/connobject.c src/connobject.c
--- mod_python-3.3.1-orig/src/connobject.c	2006-12-03 05:36:37.000000000 +0100
+++ src/connobject.c	2011-04-05 20:34:25.000000000 +0200
@@ -139,7 +139,7 @@
     bytes_read = 0;
 
     while ((bytes_read < len || len == 0) &&
-           !(b == APR_BRIGADE_SENTINEL(b) ||
+           !(b == APR_BRIGADE_SENTINEL(bb) ||
              APR_BUCKET_IS_EOS(b) || APR_BUCKET_IS_FLUSH(b))) {
 
         const char *data;
