diff -u -d -r pywebkitgtk-1.1.7/configure pywebkitgtk-1.1.7-patched/configure
--- pywebkitgtk-1.1.7/configure	2009-10-03 06:16:21.000000000 +0200
+++ pywebkitgtk-1.1.7-patched/configure	2010-07-11 21:42:14.933035488 +0200
@@ -12366,7 +12366,7 @@
 
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for pygtk codegendir" >&5
 $as_echo_n "checking for pygtk codegendir... " >&6; }
-PYGTK_CODEGENDIR=`$PKG_CONFIG --variable=codegendir pygtk-2.0`
+PYGTK_CODEGENDIR=`$PKG_CONFIG --variable=codegendir pygobject-2.0`
 
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $PYGTK_CODEGENDIR" >&5
 $as_echo "$PYGTK_CODEGENDIR" >&6; }
diff -u -d -r pywebkitgtk-1.1.7/configure.ac pywebkitgtk-1.1.7-patched/configure.ac
--- pywebkitgtk-1.1.7/configure.ac	2009-10-03 05:35:23.000000000 +0200
+++ pywebkitgtk-1.1.7-patched/configure.ac	2010-07-11 21:39:27.474455577 +0200
@@ -66,7 +66,7 @@
 
 dnl check for pygtk codegen dir
 AC_MSG_CHECKING(for pygtk codegendir)
-PYGTK_CODEGENDIR=`$PKG_CONFIG --variable=codegendir pygtk-2.0`
+PYGTK_CODEGENDIR=`$PKG_CONFIG --variable=codegendir pygobject-2.0`
 AC_SUBST(PYGTK_CODEGENDIR)
 AC_MSG_RESULT($PYGTK_CODEGENDIR)
 
