--- GPU_MD5_Crack_0.2.3_LGPL/main.cpp	2009-07-10 05:16:34.000000000 +0000
+++ GPU_MD5_Crack_0.2.3_LGPL-patched/main.cpp	2010-07-31 19:50:47.000000000 +0000
@@ -501,9 +501,9 @@
 		if(refresh_elapsed_time_ms >= REFRESH_RATE_MS)
 		{
 			elapsed_time_ms = refresh_elapsed_time_ms / counter_time;
-			printf("Progress %lld%%, " ,(i+1)*100/nb_try);
-			printf("Pwd:%s, ", pwdstr_curr);
-			printf("Instant %01.02f Mhash/s(%01.02f ms)\r",(MAX_PWD_INIT/elapsed_time_ms)/1000.0, elapsed_time_ms);
+			//printf("Progress %lld%%, " ,(i+1)*100/nb_try);
+			//printf("Pwd:%s, ", pwdstr_curr);
+			//printf("Instant %01.02f Mhash/s(%01.02f ms)\r",(MAX_PWD_INIT/elapsed_time_ms)/1000.0, elapsed_time_ms);
 			fflush(stdout);
 			refresh_elapsed_time_ms = 0;
 			counter_time = 0;
@@ -530,7 +530,7 @@
 		}
 
 	}
-	printf("End Pwd:%s                                                           \n", pwdstr_curr);
+	printf("End Pwd:%s\n", pwdstr_curr);
 	printf("Instant %01.02f Mhash/s(%01.02f ms)\n",(MAX_PWD_INIT/elapsed_time_ms)/1000.0, elapsed_time_ms);
 	printf("Average %01.02f Mhash/s, Total Time:%01.02fs(%01.02f ms)\n",(total_nb_pwd/total_elapsed_time_ms)/1000.0, total_elapsed_time_ms/1000.0, total_elapsed_time_ms);
 	printf("\nMD5 brute force finished, no password found\n");
